Add AadAuth support in configure_misp.sh

shieldsurge opened the pull request:
Add support for enabling Azure AD (aka Entra) authentication via the configure_misp.sh script. The function to enable Azure AD auth mirrors the code in the existing functions to enable OIDC and LDAP auth.

Environment variables are provided to configure available options in the AadAuth plugin: https://github.com/MISP/MISP/tree/2.4/app/Plugin/AadAuth

This commit addresses issue "AAD Auth not working" #34. https://github.com/MISP/misp-docker/issues/34

shieldsurge committed:
Add AadAuth support in configure_misp.sh

Add support for enabling Azure AD (aka Entra) authentication via the configure_misp.sh script. The function to enable Azure AD auth mirrors the code in the existing functions to enable OIDC and LDAP auth.

Environment variables are provided to configure available options in the AadAuth plugin:
https://github.com/MISP/MISP/tree/2.4/app/Plugin/AadAuth

This commit addresses issue "AAD Auth not working" #34.
https://github.com/MISP/misp-docker/issues/34

shieldsurge committed:
Update configure_misp.sh

Remove accidentally leftover echo debugging statement

shieldsurge committed:
Update configure_misp.sh - Add string quotes to default AAD variable values

Add string quotes to default AAD variable values as a best practice.

ostefano commented on core/files/configure_misp.sh:
```diff
@@ -157,6 +157,60 @@ set_up_ldap() {
+    check_env_vars AAD_CLIENT_ID AAD_TENANT_ID AAD_CLIENT_SECRET BASE_URL
+
+    # Configure unset optional AAD environment variables to default values
+    [ -z "$AAD_REDIRECT_URI" ] && AAD_REDIRECT_URI="${BASE_URL}/users/login"
```
Other auth mechanisms do not set default but leave suggested values in the `template.env` file. AAD should do the same since MS is not the only implementation (let me know if that is instead the case).

shieldsurge replied:
The goal here was to provide a few default variable values to allow the administrator to specify a minimal set of environment variables for AAD auth while still preserving their ability to override any of the default variables by setting the environment variables explicitly.

For example, the [AadAuth plugin readme](https://github.com/MISP/MISP/blob/2.4/app/Plugin/AadAuth/README.md) provides example values for the `auth_provider` URI and the `auth_provider_user` URI, and comments "No change required" for both. I believe these values are the same for everybody, so the idea was to set them to the default values unless explicitly overridden in environment variables.

There's also optional functionality in the AadAuth plugin to have MISP require the authenticating end user to be a member of one of three administrator-defined Azure AD groups to allow login. This functionality isn't very useful at the moment because there's no mechanism to map roles in MISP to Azure AD groups as far as I can tell. Therefore, I made the default value for this functionality off (`AAD_CHECK_GROUPS=false`), and the three Azure AD group names are just set to the example values specified in the AadAuth plugin readme page.

`AAD_REDIRECT_URI` is going to be `"${BASE_URL}/users/login"` in 99% of implementations, so I specified it as the default which can be overridden by setting the env variable.

All that being said, I don't feel strongly about this issue. If the desired practice is to require the administrator to explicitly specify all possible AAD environment variables, just let me know and I'll adjust the code accordingly.

@ostefano for awareness.

ostefano commented on core/files/configure_misp.sh:
```diff
@@ -157,6 +157,60 @@ set_up_ldap() {
+    fi
+
+    # Check required variables
+    check_env_vars AAD_CLIENT_ID AAD_TENANT_ID AAD_CLIENT_SECRET BASE_URL
```
No need to check BASE_URL, I would rather check all other vars

shieldsurge replied:
I omitted some of the environment variables from the `check_env_vars` call because I set them to default values if not specified, per the discussion below. If that's not the direction we want to go in, then I will add those variables to the `check_env_vars` call to require them to be set explicitly.

ostefano reviewed (changes requested):
Thank you for the PR.

We should also update `docker-compose.yml` and `template.env`

shieldsurge committed:
Update docker-compose.yml - Add env variables for AAD auth

shieldsurge committed:
Update template.env - Provide env variable examples for AAD auth

Specify environment variables for enabling Azure AD (Entra) authentication and provide example values where applicable.

shieldsurge commented:
> Thank you for the PR.
> 
> We should also update `docker-compose.yml` and `template.env`

`docker-compose.yml` and `template.env` have been updated.

ostefano commented:
> > Thank you for the PR.
> > We should also update `docker-compose.yml` and `template.env`
> 
> `docker-compose.yml` and `template.env` have been updated.

Thank you,  but defaults are still set in the configure_misp file. Can we remove it from there so all auth mechanisms are consistent with each other?

shieldsurge committed:
Update configure_misp.sh - Require all AAD env vars

Remove code which allowed some AAD variables to be optional and set to default values. Require all AAD environment variables to be explicitly specified if AAD is enabled consistent with approach used by other auth mechanisms (OIDC and LDAP).

shieldsurge commented:
> > > Thank you for the PR.
> > > We should also update `docker-compose.yml` and `template.env`
> > 
> > 
> > `docker-compose.yml` and `template.env` have been updated.
> 
> Thank you, but defaults are still set in the configure_misp file. Can we remove it from there so all auth mechanisms are consistent with each other?

Understood, assuming you saw my rationale for default values in the review section. I went ahead and removed the default values from `core/files/configure_misp.sh`. It now requires all AAD env vars to be explicitly set when using AAD to be consistent with the approach used by the other auth mechanisms.

ostefano commented:
> > > > Thank you for the PR.
> > > > We should also update `docker-compose.yml` and `template.env`
> > > 
> > > 
> > > `docker-compose.yml` and `template.env` have been updated.
> > 
> > 
> > Thank you, but defaults are still set in the configure_misp file. Can we remove it from there so all auth mechanisms are consistent with each other?
> 
> Understood, assuming you saw my rationale for default values in the review section. I went ahead and removed the default values from `core/files/configure_misp.sh`. It now requires all AAD env vars to be explicitly set when using AAD to be consistent with the approach used by the other auth mechanisms.

Mmm, I don't see any reply in the review section. Did I miss your comments?

shieldsurge commented:
> > > > > Thank you for the PR.
> > > > > We should also update `docker-compose.yml` and `template.env`
> > > > 
> > > > 
> > > > `docker-compose.yml` and `template.env` have been updated.
> > > 
> > > 
> > > Thank you, but defaults are still set in the configure_misp file. Can we remove it from there so all auth mechanisms are consistent with each other?
> > 
> > 
> > Understood, assuming you saw my rationale for default values in the review section. I went ahead and removed the default values from `core/files/configure_misp.sh`. It now requires all AAD env vars to be explicitly set when using AAD to be consistent with the approach used by the other auth mechanisms.
> 
> Mmm, I don't see any reply in the review section. Did I miss your comments?

I can't seem to figure out how to link to the comment, but here's what I wrote:

The goal here was to provide a few default variable values to allow the administrator to specify a minimal set of environment variables for AAD auth while still preserving their ability to override any of the default variables by setting the environment variables explicitly.

For example, the [AadAuth plugin readme](https://github.com/MISP/MISP/blob/2.4/app/Plugin/AadAuth/README.md) provides example values for the `auth_provider` URI and the `auth_provider_user` URI, and comments "No change required" for both. I believe these values are the same for everybody, so the idea was to set them to the default values unless explicitly overridden in environment variables.

There's also optional functionality in the AadAuth plugin to have MISP require the authenticating end user to be a member of one of three administrator-defined Azure AD groups to allow login. This functionality isn't very useful at the moment because there's no mechanism to map roles in MISP to Azure AD groups as far as I can tell. Therefore, I made the default value for this functionality off (`AAD_CHECK_GROUPS=false`), and the three Azure AD group names are just set to the example values specified in the AadAuth plugin readme page.

`AAD_REDIRECT_URI` is going to be `"${BASE_URL}/users/login"` in 99% of implementations, so I specified it as the default which can be overridden by setting the env variable.

All that being said, I don't feel strongly about this issue. If the desired practice is to require the administrator to explicitly specify all possible AAD environment variables, just let me know and I'll adjust the code accordingly.

shieldsurge commented:
> Mmm, I don't see any reply in the review section. Did I miss your comments?

It was because I never submitted the review 🤦
You should be able to see the comments now.

ostefano reviewed (approved):
I will merge this, but add some changes afterwards
